### PR TITLE
Add a value judgement for the cloud in controllermanager.go

### PR DESCRIPTION
**Pull request by ping035627**
The PR add a value judgement for the cloud, because cloudprovider.InitCloudProvider maybe return nil for the cloud.

### CODE REVIEW

**Comment by jdef** [comment]
@k8s-bot ok to test

**Comment by jdef** [comment]
@ping035627 any reason that you didn't just copy/paste the code block from here? https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L258

**Comment by ping035627** [comment]
@jdef
 I  just  think  the  style  is better  than  that according  to the go style  advises.  but  now  I  think  here shouldn' t  use glog.Fatal,  and  I  will  modify  it  soon. what' s  your advises? thanks.

} else if routes, ok := cloud.Routes(); !ok {
	glog.Warning("configure-cloud-routes is set, but cloud provider does not support routes. Will not configure cloud provider routes.")
} else {

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

**Comment by ping035627** [comment]
@jdef I've modify, please review again, thanks.

File: contrib/mesos/pkg/controllermanager/controllermanager.go [hunk 1]
```diff
@@ -172,12 +172,16 @@ func (s *CMServer) Run(_ []string) error {
...
-		routes, ok := cloud.Routes()
-		if !ok {
-			glog.Fatal("Cloud provider must support routes if configure-cloud-routes is set")
+		if cloud == nil {
+			glog.Warning("configure-cloud-routes is set, but no cloud provider specified. Will not configure cloud provider routes.")
+		} else if routes, ok := cloud.Routes(); !ok {
+			glog.Warning("configure-cloud-routes is set, but cloud provider does not support routes. Will not configure cloud provider routes.")
+		} else {
+			routeController := routecontroller.New(routes, clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "route-controller")), s.ClusterName, clusterCIDR)
+			routeController.Run(s.NodeSyncPeriod.Duration)
```
Comment (jdef) [inline contrib/mesos/pkg/controllermanager/controllermanager.go]: missing the `time.Sleep` after this line that's present upstream

File: contrib/mesos/pkg/controllermanager/controllermanager.go [hunk 2]
```diff
@@ -172,12 +172,16 @@ func (s *CMServer) Run(_ []string) error {
...
 		}
-		routeController := routecontroller.New(routes, clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "route-controller")), s.ClusterName, clusterCIDR)
-		routeController.Run(s.NodeSyncPeriod.Duration)
+	} else {
```
Comment (jdef) [inline contrib/mesos/pkg/controllermanager/controllermanager.go]: why not just copy/paste the else conditions from the upstream code. looks like you just combined the error messages here.

**Comment by jdef** [comment]
thanks for spending time fixing this up. looking better. the reason that i'd rather just have the code copy/pasted from upstream is that it's easier to maintain over time. otherwise differences between the two impls grow to a point where it becomes hard to see subtle implementation changes at a glance.

File: contrib/mesos/pkg/controllermanager/controllermanager.go (hunk 2 above)
Comment (ping035627) [inline contrib/mesos/pkg/controllermanager/controllermanager.go]: yes, I think it' s better  to combine the error message,  and  it seem  more concise.

**Comment by ping035627** [comment]
ok, I understand,  and  I will copy/pasted  the  code from upstream soon.

**Comment by ping035627** [comment]
I've modified two places and passed the test, please review again, thanks.
1) add time.Sleep, copy/pasted the code from upstream;
2) add a error info to log. (the same to upstream)

About the combined error messages, it have merged into upstream.
https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L268

**Comment by jdef** [comment]
Thanks! lgtm

**Pull request merged**